x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/FrozenCacheService.java

```diff
@@ -754,89 +681,137 @@ protected void alreadyClosed() {
+         */
         public StepListener<Integer> populateAndRead(
             final ByteRange rangeToWrite,
             final ByteRange rangeToRead,
             final RangeAvailableHandler reader,
             final RangeMissingHandler writer,
             final Executor executor
         ) {
+            assert rangeToWrite.length() > 0 : "should not try to populate and read empty range";
             StepListener<Integer> stepListener = null;
-            final long writeStart = rangeToWrite.start();
-            final long readStart = rangeToRead.start();
-            for (int region = getRegion(rangeToWrite.start()); region <= getEndingRegion(rangeToWrite.end()); region++) {
-                final ByteRange subRangeToWrite = mapSubRangeToRegion(rangeToWrite, region);
-                final ByteRange subRangeToRead = mapSubRangeToRegion(rangeToRead, region);
-                if (subRangeToRead.length() == 0L) {
-                    // nothing to read, skip
-                    if (stepListener == null) {
-                        stepListener = new StepListener<>();
-                        stepListener.onResponse(0);
-                    }
-                    continue;
+            final int lastRegion = regionForPosition(rangeToWrite.end() - 1);
+            final int firstRegion = regionForPosition(rangeToWrite.start());
+            final PopulateAndReadContext context = new PopulateAndReadContext();
+            for (int region = firstRegion; region <= lastRegion; region++) {
+                final CacheFileRegion fileRegion;
+                try {
+                    fileRegion = get(cacheKey, fileSize, region);
+                } catch (Exception e) {
+                    // if we can't acquire enough cache pages for all the regions in a write we fail the execution right away so that the
+                    // caller can move on to a direct read bypassing the cache
+                    context.fail(e);
+                    throw e;
                 }
-                final CacheFileRegion fileRegion = get(cacheKey, length, region);
-                final long regionStart = getRegionStart(region);
-                final long writeOffset = writeStart - regionStart;
-                final long readOffset = readStart - regionStart;
-                final StepListener<Integer> lis = fileRegion.populateAndRead(
+                final StepListener<Integer> listener = new StepListener<>();
+
+                // get the channel to the cache page that the current file region has acquired
+                final SharedBytes.IO fileChannel = sharedBytes.getFileChannel(fileRegion.sharedPageIndex);
+                final Releasable decrementRef = Releasables.releaseOnce(() -> Releasables.close(fileChannel::decRef, fileRegion::decRef));
+                // as we are potentially going to write to the region and its IO we acquire another reference for both the region
+                // and the IO for writing
+                fileChannel.incRef();
```

Comment:
Can we do this only when we know we'll effectively have some gap to write?

Reply:
We can't unfortunately. We decrement the ref count when there's no gap in what we try to read via the listener. But we could still have a gap in our write range. I couldn't find a way to make this selective without more changes so this felt easiest for now. We could technically do something like add the listener callback after we are done with writes to only have one increment change overall but that felt even more confusing than this to me at least.